### PR TITLE
Fix #30557 Narrow non-declared unions by discriminant

### LINKED ISSUE #30557: Unknown types narrowed to an intersection won't narrow further
This fails [on the playground](https://www.typescriptlang.org/play/#src=interface%20TypeA%20%7B%0A%09Name%3A%20%22TypeA%22%3B%0A%09Value1%3A%20%22Cool%20stuff!%22%3B%0A%7D%0A%0Ainterface%20TypeB%20%7B%0A%09Name%3A%20%22TypeB%22%3B%0A%09Value2%3A%200%3B%0A%7D%0A%0Atype%20Type%20%3D%20TypeA%20%7C%20TypeB%3B%0A%0Adeclare%20function%20isType(x%3A%20unknown)%3A%20x%20is%20Type%3B%0A%0Afunction%20WorksProperly(data%3A%20Type)%20%7B%0A%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%2F%2F%20data%3A%20TypeA%0A%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%7D%0A%7D%0A%0Afunction%20DoesNotWork(data%3A%20unknown)%20%7B%0A%09if%20(isType(data))%20%7B%0A%09%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%09%2F%2F%20data%3A%20Type%0A%09%09%09%2F%2F%20data%20should%20be%20TypeA%0A%09%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%09%7D%0A%09%7D%0A%7D%0A), and I believe it also fails on typescript@3.4.0-dev.20190323

**TypeScript Version:**  typescript@3.4.0-dev.20190323

**Search Terms:** unknown intersection failure types narrow

**Code**

```ts
interface TypeA {
	Name: "TypeA";
	Value1: "Cool stuff!";
}

interface TypeB {
	Name: "TypeB";
	Value2: 0;
}

type Type = TypeA | TypeB;

declare function isType(x: unknown): x is Type;

function WorksProperly(data: Type) {
	if (data.Name === "TypeA") {
		// data: TypeA
		const value1 = data.Value1;
	}
}

function DoesNotWork(data: unknown) {
	if (isType(data)) {
		if (data.Name === "TypeA") {
			// data: Type
			// data should be TypeA
			const value1 = data.Value1; // type error!
		}
	}
}
```

**Expected behavior:** in `DoesNotWork`, `data` should be TypeA after we can guarentee it is named "TypeA"

**Actual behavior:** `data` is still only a `Type`

[**Playground Link:** Here](https://www.typescriptlang.org/play/#src=interface%20TypeA%20%7B%0A%09Name%3A%20%22TypeA%22%3B%0A%09Value1%3A%20%22Cool%20stuff!%22%3B%0A%7D%0A%0Ainterface%20TypeB%20%7B%0A%09Name%3A%20%22TypeB%22%3B%0A%09Value2%3A%200%3B%0A%7D%0A%0Atype%20Type%20%3D%20TypeA%20%7C%20TypeB%3B%0A%0Adeclare%20function%20isType(x%3A%20unknown)%3A%20x%20is%20Type%3B%0A%0Afunction%20WorksProperly(data%3A%20Type)%20%7B%0A%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%2F%2F%20data%3A%20TypeA%0A%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%7D%0A%7D%0A%0Afunction%20DoesNotWork(data%3A%20unknown)%20%7B%0A%09if%20(isType(data))%20%7B%0A%09%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%09%2F%2F%20data%3A%20Type%0A%09%09%09%2F%2F%20data%20should%20be%20TypeA%0A%09%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%09%7D%0A%09%7D%0A%7D%0A)

**Related Issues:**  I looked through several pages of issues and did not find anything similar.

### PR DESCRIPTION
**Pull request by jack-williams**
Fixes #30557

### CODE REVIEW

**Comment by Validark** [comment]
Short and seemingly straightforward. Hopefully someone with the power to merge this can review it soon. Thank you for tracking this bug down and fixing it!

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this

**Comment by jack-williams** [comment]
Perf regression?

**Comment by jack-williams** [comment]
For large unions without a discriminant I was computing the union of the property twice, which I guess is expensive. Not sure if this is enough to fix the problem however.

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this new commit

**Comment by jack-williams** [comment]
I dont know if anyone with more knowledge than me sees something obvious that I’m missing. One alternative might be to limit the narrowing to cases where the declared type is `unknown` or `object`, though that doesn’t feel especially principled.

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this because maybe it works now!

**Comment by jack-williams** [comment]
Sorry @typescript-bot, I hope I don't let you down

**Comment by Validark** [comment]
Psyched!

**Comment by jack-williams** [comment]
@RyanCavanaugh 

The RWC tests were running at 40m and going, so I don't think my change fixed that. I've changed it to only ever narrow from one type. Before, if the declared type was not a union but narrowed to a big union, then each subsequent narrowing would hit a different property cache.

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this

**Comment by RyanCavanaugh** [comment]
Forgot to sync this branch with master; need to rerun

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this

**Comment by RyanCavanaugh** [comment]
RWC is actually clean

@typescript-bot perf test this

**Comment by jack-williams** [comment]
So it seems like the catastrophic perf regression I introduced is fixed. Is the delta column the one we should be looking at?

**Comment by DanielRosenwasser** [comment]
@RyanCavanaugh can you check if the following code works? The set-once nature of the fix feels a bit suspicious.

```ts
type Foo = { kind: "a", a: number } | { kind: "b", b: number };
type Bar = { kind: "c", d: number } | { kind: "d", d: number };

function isFoo(x: unknown): x is Foo;
function isBar(x: unknown): x is Bar;

function blah(x: unknown) {
    if (isFoo(x)) {
        if (x.kind === "a") {
            let a = x.a;
        }
        else if (x.kind === "b") {
            let b = x.b;
        }
    }
    else if (isBar(x)) {
        if (x.kind === "c") {
            let c = x.c;
        }
        else if (x.kind === "d") {
            let d = x.d;
        }
    }
}
```

**Comment by jack-williams** [comment]
Ye I remember a scenario like that crossing my mind when I wrote the fix. I did some experimentation but I’m not sure I tried that exact one - there was a point where stuff was working that I didn’t expect to. In the end I was just attempting to restore perf - once that’s reached the actual solution should be scrutinised

**Comment by jack-williams** [comment]
@DanielRosenwasser @RyanCavanaugh I checked your code on my branch and it seems to work correctly - but I must admit I completely suspicious of this. It definitely needs to be reviewed by someone with a decent grounding in the CFA.

**Comment by DanielRosenwasser** [comment]
@RyanCavanaugh and @ahejlsberg were able to find a strange case where things don't work out right - I think @RyanCavanaugh can provide the sample.

**Comment by jack-williams** [comment]
I think the approach that I had was definitely wrong. I've changed it slightly to remove the set once behaviour, and instead set and unset at the point types get narrowed to something discriminable. It's likely that's its not quite right too, but I just want to put something out there to test.

BTW, if someone on the team knows how to do this properly and wants to scoop me, please go ahead. Just let me know if someone is working on it.

**Comment by DanielRosenwasser** [comment]
Just as a heads up, your last commit doesn't seem to be associated with your GitHub account. You can either make sure your GitHub account is associated with the email address you're using for your commits, or rebase and amend your commits to fix the author name and email.

**Comment by RyanCavanaugh** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
Is the bind time where we would expect the perf hit here?

**Comment by RyanCavanaugh** [comment]
We had a binder regression a while back that was unrelated to this change

**Comment by RyanCavanaugh** [comment]
@typescript-bot perf test this
@typescript-bot test this
@typescript-bot user test this

**Comment by jack-williams** [comment]
Should I merge the bot PR into my branch? Looks like some of the changes might be due to #32904.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17271,6 +17272,10 @@ namespace ts {
                 return result;
             }
 
+            function isFlowToRootDiscriminant(type: Type, candidate: Type) {
+                return (type.flags & TypeFlags.Unknown | TypeFlags.Any) && (candidate.flags & TypeFlags.Union) !== 0;
```
Comment (jack-williams) [inline src/compiler/checker.ts]: Might want to extend this to include `object`.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -17271,6 +17272,10 @@ namespace ts {
                 return result;
             }
 
+            function isFlowToRootDiscriminant(type: Type, candidate: Type) {
```
Comment (jack-williams) [inline src/compiler/checker.ts]: Change to type predicate?

File: src/compiler/checker.ts (hunk 1 above)
Comment (jack-williams) [inline src/compiler/checker.ts]: Precedence looks completely wrong.

**Comment by RyanCavanaugh** [comment]
@ahejlsberg thoughts on this? Perf results look solid and this is a pretty frequent source of user reports

**Comment by jack-williams** [comment]
@RyanCavanaugh You might want to re-test perf. I think the precedence of the conditional I wrote was slightly wrong (changed now), so I just want to make sure that I wasn't excluding a load of cases.

**Comment by ahejlsberg** [comment]
The approach in this PR unfortunately doesn't work in composite scenarios:

```ts
declare function isTypeAB(x: unknown): x is { kind1: 'a', a: 1 } | { kind1: 'b', b: 2 };
declare function isTypeCD(x: unknown): x is { kind2: 'c', c: 3 } | { kind2: 'd', d: 4 };

function test(x: unknown) {
    if (isTypeAB(x)) {
        if (x.kind1 === 'a') {
            x.a;  // Ok
        }
    }
    if (isTypeCD(x)) {
        if (x.kind2 === 'c') {
            x.c;  // Ok
        }
    }
    if (isTypeAB(x)) {
        if (isTypeCD(x)) {
            if (x.kind1 === 'a') {
                x.a;  // Ok
            }
            if (x.kind2 === 'c') {
                x.c;  // Error
            }
        }
    }
}
```

The issue is that `rootDiscriminable` only caches the effects of a single type predicate when in reality there could be multiple active predicates.

**Comment by jack-williams** [comment]
Does the flaw lead to soundness issues, or just incompleteness?

**Comment by RyanCavanaugh** [comment]
> You might want to re-test perf. 

@jack-williams I added you as a collaborator so you now have full access for typescript-bot commands. Enjoy!

**Comment by jack-williams** [comment]
@RyanCavanaugh Thank you!

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@RyanCavanaugh @ahejlsberg

I've implemented it slightly differently, and in a way that should work in the composite case.

Perf doesn't look disastrous, though I'm not quite sure how to interpret the tables fully. The worst delta in check time is +1.07%, is that too much?

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this
@typescript-bot test this

**Comment by jack-williams** [comment]
Looks like using the declared type (if valid) is slightly better than using the currently contained union, but it's hard to know if the deltas are significant.

**Comment by RyanCavanaugh** [comment]
Sample code for discussion
```ts
type A = { x: number };
type B = A & { kind: "B"; y: number };
type C = A & { kind: "C"; z: number };

function isBorC(a: A): a is B | C {
    return (a as any).kind === "B" || (a as any).kind === "C";
}

// "Correct" signature, doesn't work
function fn1(a: A) {
    if (isBorC(a)) {
        if (a.kind === "B") {
            // Incorrect error
            a.y;
        }
    }
}

// Nonsense equivalent signature, works
function fn2(a: A | B | C) {
    if (isBorC(a)) {
        if (a.kind === "B") {
            // OK
            a.y;
        }
    }
}
```

**Comment by jack-williams** [comment]
Both functions are OK with the currently implemented solution in this PR. Should they behave differently?

**Comment by RyanCavanaugh** [comment]
@jack-williams the examples were just for demonstrating what is wrong with the current behavior since we were discussing this today at the design meeting

**Comment by jack-williams** [comment]
@RyanCavanaugh Ah, my bad! Thought I overlooked something silly.

**Comment by jack-williams** [comment]
Will resolve conflicts this week.

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this now I have removed the union cache

**Comment by jack-williams** [comment]
@typescript-bot perf test this now I've added cases for assignment

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
@typescript-bot perf test this with recent CFA changes

**Pull request closed without merging**